osc/util/models.py

```diff
@@ -390,3 +409,362 @@ def dict(self):
+
+                kwargs[field_name] = values
+                continue
+
+            if field.is_model:
+                # field contains an instance of XmlModel
+                assert xml_name is not None
+                node = root.find(xml_name)
+                if node is None:
+                    continue
+                klass = field.origin_type
+                kwargs[field_name] = klass.from_xml(node)
+
+                # clear node as it was checked in from_xml() already
+                node.text = None
+                node.attrib = {}
+                node[:] = []
+
+                cls._remove_empty_node(root, node)
+                continue
+
+            # field contains a scalar
+            node = root.find(xml_name)
+            if node is None:
+                continue
+            value = cls.value_from_string(field, node.text)
+            node.text = None
+            if value is None:
+                if field.is_optional:
+                    continue
+                value = ""
+            kwargs[field_name] = value
+            cls._remove_empty_node(root, node)
+
+        cls._remove_empty_node(None, root)
+
+        obj = cls(**kwargs)
+        obj.__dict__["_root"] = orig_root
+        return obj
+
+    @classmethod
+    def xml_request(cls, method: str, apiurl: str, path: List[str], query: Optional[dict] = None, data: Optional[str] = None) -> urllib3.response.HTTPResponse:
+        from ..connection import http_request
+        from ..core import makeurl
+        url = makeurl(apiurl, path, query)
+        # TODO: catch HTTPError and return the wrapped response as XmlModel instance
+        return http_request(method, url, data=data, retry_on_400=False)
+
+    def do_update(self, other: "XmlModel") -> None:
+        """
+        Update values of the fields in the current model instance from another.
+        """
+        self._values = copy.deepcopy(other._values)
+
+    def do_edit(self) -> Tuple[str, str, "XmlModel"]:
+        """
+        Serialize model as XML and open it in an editor for editing.
+        Return a tuple with:
+          * a string with original data
+          * a string with edited data
+          * an instance of the class with edited data loaded
+
+        IMPORTANT: This method is always interactive.
+        """
+        from ..core import run_editor
+        from ..output import get_user_input
+
+        def write_file(f, data):
+            f.seek(0)
+            f.write(data)
+            f.truncate()
+            f.flush()
+
+        with tempfile.NamedTemporaryFile(mode="w+", encoding="utf-8", prefix="obs_xml_", suffix=".xml") as f:
+            original_data = self.to_string()
+            write_file(f, original_data)
+
+            while True:
+                run_editor(f.name)
+                try:
+                    edited_obj = self.__class__.from_file(f.name)
+                    f.seek(0)
+                    edited_data = f.read()
```

Comment:
Shouldn't we truncate here too?

Reply:
No, because we truncate on `write()` while this performs `read()`